sql: corruption of `SetupFlowRequest`

pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test.TestSqlLiteLogic_testindexcommute10slt_good_5_test [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SQLiteLogicTestsBazel/20862317?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SQLiteLogicTestsBazel/20862317?buildTab=artifacts#/) on release-25.2.10-rc @ [a2cf0e12d277179df853e7e494c0155a6839edf3](https://github.com/cockroachdb/cockroach/commits/a2cf0e12d277179df853e7e494c0155a6839edf3):


```
[16:39:01] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 6686 statements
[16:39:03] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 7008 statements
[16:39:05] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 7267 statements
[16:39:07] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 7585 statements
[16:39:09] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 7924 statements
[16:39:11] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 8225 statements
[16:39:13] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 8526 statements
[16:39:15] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 8823 statements
[16:39:17] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 9145 statements
[16:39:19] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 9449 statements
[16:39:21] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test: 9777 statements
[16:39:22] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3014/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/sqlitelogictest/tests/fakedist-disk/fakedist-disk_test_/fakedist-disk_test.runfiles/com_github_cockroachdb_cockroach/external/com_github_cockroachdb_sqllogictest/test/index/commute/10/slt_good_5.test with config fakedist-disk: 10032 tests, 0 failures
panic: runtime error: slice bounds out of range [-7:] [recovered]
	panic: runtime error: slice bounds out of range [-7:]

goroutine 492 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00ff4e848?, {0x7049950, 0xc00257e140})
	pkg/util/stop/stopper.go:227 +0x65
panic({0x5f6f600?, 0xc00dce9b48?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*FlowSpec).MarshalToSizedBuffer(0xc00ff4e848?, {0xc009a24000, 0x69f, 0xa04?})
	bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:833 +0x1a7
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*SetupFlowRequest).MarshalToSizedBuffer(0xc00ff4e800, {0xc009a24000, 0xa04, 0xa04})
	bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:781 +0x5d2
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*SetupFlowRequest).Marshal(0xc00ff4e800)
	bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:690 +0x4d
github.com/cockroachdb/cockroach/pkg/rpc.codec.Marshal({}, {0x6085dc0?, 0xc00ff4e800})
	pkg/rpc/codec.go:29 +0x67
google.golang.org/grpc.encode({0x7f53e9af3ed0?, 0xc000f0ee80?}, {0x6085dc0?, 0xc00ff4e800?})
	external/org_golang_google_grpc/rpc_util.go:633 +0x3e
google.golang.org/grpc.prepareMsg({0x6085dc0?, 0xc00ff4e800?}, {0x7f53e9af3ed0?, 0xc000f0ee80?}, {0x0, 0x0}, {0x702cd70, 0xaffec80})
	external/org_golang_google_grpc/stream.go:1766 +0xc7
google.golang.org/grpc.(*clientStream).SendMsg(0xc00a415680, {0x6085dc0, 0xc00ff4e800})
	external/org_golang_google_grpc/stream.go:882 +0xd7
google.golang.org/grpc.invoke({0x7046820?, 0xc00ccbc5a0?}, {0x6404614?, 0x41d9b0?}, {0x6085dc0, 0xc00ff4e800}, {0x6085780, 0xc010296eb0}, 0x429dc5?, {0xc003244440, ...})
	external/org_golang_google_grpc/call.go:75 +0x9f
github.com/cockroachdb/cockroach/pkg/rpc.NewContext.ClientInterceptor.func8({0x7046820, 0xc00ccbc5a0}, {0x6404614, 0x2b}, {0x6085dc0, 0xc00ff4e800}, {0x6085780, 0xc010296eb0}, 0xc008665808, 0x66fb340, ...)
	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:240 +0x3d2
google.golang.org/grpc.(*ClientConn).Invoke(0xc008665808, {0x7046820, 0xc00ccbc5a0}, {0x6404614, 0x2b}, {0x6085dc0, 0xc00ff4e800}, {0x6085780, 0xc010296eb0}, {0x0, ...})
	external/org_golang_google_grpc/call.go:40 +0x229
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.(*distSQLClient).SetupFlow(0xc0137efd88, {0x7046820, 0xc00ccbc5a0}, 0xc00ff4e800, {0x0, 0x0, 0x0})
	bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:511 +0xc5
github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run({{0x7046820, 0xc00ccbc5a0}, 0xc000fe4cc0, 0xc00ff4e800, 0x2, 0xc00c08ea80})
	pkg/sql/distsql_running.go:133 +0x1ce
github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0x0?, 0x7049950?})
	pkg/sql/distsql_running.go:166 +0x77
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x632f848, 0xe})
	pkg/util/stop/stopper.go:500 +0x1e2
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 29
	pkg/util/stop/stopper.go:489 +0x466
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSqlLiteLogic_testindexcommute10slt_good_5_test.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57921

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -568,10 +568,14 @@ func (dsp *DistSQLPlanner) setupFlows(
 	var runnerSpan *tracing.Span
 	// This span is necessary because it can outlive its parent.
 	runnerCtx, runnerSpan = tracing.ChildSpan(runnerCtx, "setup-flow-async" /* opName */)
+	// runnerDone will be closed whenever all concurrent SetupFlowRequest RPCs
+	// are complete.
+	runnerDone := make(chan struct{})
 	// runnerCleanup can only be executed _after_ all issued RPCs are complete.
 	runnerCleanup := func() {
 		cancelRunnerCtx()
 		runnerSpan.Finish()
+		close(runnerDone)
 	}
 	// Make sure that we call runnerCleanup unless a new goroutine takes that
 	// responsibility.
@@ -642,24 +646,20 @@ func (dsp *DistSQLPlanner) setupFlows(
 	// setup of a remote flow fails, we want to eagerly cancel all the flows,
 	// and we do so in a separate goroutine.
 	//
-	// We need to synchronize the new goroutine with flow.Cleanup() being called
-	// since flow.Cleanup() is the last thing before DistSQLPlanner.Run returns
-	// at which point the rowResultWriter is no longer protected by the mutex of
-	// the DistSQLReceiver.
-	// TODO(yuzefovich): think through this comment - we no longer have mutex
-	// protection in place, so perhaps this can be simplified.
-	cleanupCalledMu := struct {
-		syncutil.Mutex
-		called bool
-	}{}
+	// We need to ensure that the local flow Cleanup() is blocked until all RPCs
+	// are performed. This is needed since we release the PhysicalInfrastructure
+	// back into the sync.Pool, and FlowSpecs for concurrent RPCs might
+	// reference that infra.
+	var cleanupCalled atomic.Bool
 	flow.AddOnCleanupStart(func() {
-		cleanupCalledMu.Lock()
-		defer cleanupCalledMu.Unlock()
-		cleanupCalledMu.called = true
-		// Cancel any outstanding RPCs while holding the lock to protect from
-		// the context canceled error (the result of the RPC) being set on the
-		// DistSQLReceiver by the listener goroutine below.
+		// Indicate that the local flow Cleanup has started - at this point we
+		// can ignore any errors communicated by the listener goroutine below.
+		cleanupCalled.Store(true)
+		// Cancel any outstanding RPCs.
 		cancelRunnerCtx()
+		// Now block the cleanup of the local flow until all concurrent RPCs are
+		// complete.
+		<-runnerDone
 	})
 	err = dsp.stopper.RunAsyncTask(origCtx, "distsql-remote-flows-setup-listener", func(ctx context.Context) {
 		// Note that in the loop below we always receive from the result channel
@@ -673,21 +673,17 @@ func (dsp *DistSQLPlanner) setupFlows(
 			if res.err != nil && !seenError {
 				// The setup of at least one remote flow failed.
 				seenError = true
-				func() {
-					cleanupCalledMu.Lock()
-					defer cleanupCalledMu.Unlock()
-					if cleanupCalledMu.called {
-						// Cleanup of the local flow has already been performed,
-						// so there is nothing to do.
-						return
-					}
-					// First, we send the error to the DistSQL receiver to be
-					// returned to the client eventually (which will happen on
-					// the next Push or PushBatch call).
-					recv.concurrentErrorCh <- res.err
-					// Now explicitly cancel the local flow.
-					flow.Cancel()
-				}()
+				if cleanupCalled.Load() {
+					// Cleanup of the local flow has already started, so there
+					// is nothing to communicate to the local flow.
+					continue
+				}
+				// First, we send the error to the DistSQL receiver to be
+				// returned to the client eventually (which will happen on the
+				// next Push or PushBatch call).
+				recv.concurrentErrorCh <- res.err
+				// Now explicitly cancel the local flow.
+				flow.Cancel()
 			}
 		}
 	})

--- a/pkg/sql/physicalplan/physical_plan.go
+++ b/pkg/sql/physicalplan/physical_plan.go
@@ -193,8 +193,9 @@ func NewPhysicalInfrastructure(
 
 // Release resets the object and puts it back into the pool for reuse.
 func (p *PhysicalInfrastructure) Release() {
-	// We only need to nil out the local processors since these are the only
-	// pointer types.
+	for i := range p.Processors {
+		p.Processors[i] = Processor{}
+	}
 	for i := range p.LocalProcessors {
 		p.LocalProcessors[i] = nil
 	}

